[red-knot] Instance attributes: type inference clarifications (#15512)
astral-sh/ruff: crates/red_knot_python_semantic/resources/mdtest/attributes.md

# mypy shows no error here, but pyright raises "reportAttributeAccessIssue"
C.pure_instance_variable1 = "overwritten on class"

c_instance.pure_instance_variable4 = False

# TODO: After this assignment to the attribute within this scope, we may eventually want to
# narrow the `bool` type (see above) for this instance variable to `Literal[False]` here.
# This is unsound in general (we don't know what else happened to `c_instance` between the

> # assignment and the use here), but pyright supports this, presumably to provide features

Reviewer: ```suggestion
# assignment and the use here), but pyright and mypy support this, presumably to provide features
```

I can't remember where pyre ended up on this; @carljm can refresh my memory here :-) I believe they started off being very strict about soundness here, but it resulted in a lot of user confusion and not a significant increase in type safety, so then loosened their rules somewhat. But I don't know exactly where they ended up at the end of the day. When we were chatting to the pyre developers at PyCon, my recollection is that they said they regretted ever going down that road, because users had such a strong expectation of type narrowing in code like this.

Reviewer 2: I think (and Pyre playground testing just now confirmed) Pyre does the narrowing but then reverts it after an expression that seems likely to execute arbitrary code elsewhere (e.g. a call expression). This is also very much a heuristic, and one that results in behavior that is surprising to users ("why did adding this call that doesn't involve variable `c_instance` suddenly change the inferred type of `c_instance.pure_instance_variable4`??"), so I'm not particularly fond of that approach either.

I agree that we will probably just want to follow mypy and pyright here.

Author: @AlexWaygood In which sense does mypy support this? I don't see a narrowed type here: https://mypy-play.net/?mypy=latest&python=3.13&gist=4cc6d2748f364c614965b72becf538e7

Do I need to use an example that does not involve literals?

Author: > Do I need to use an example that does not involve literals?

Yes. Yes, I do. For unions, it works.

Reviewer 2: Yes, mypy is very sparing in its willingness to infer literals. If you try an optional type like `int | None` for an attribute, then assign `3` to it, mypy will narrow to `int`.

Reviewer: Yeah, mypy never narrows instance types to literal types, but you can see that it narrows the types of attributes in exactly the same way it allows you narrow the types of local variables https://mypy-play.net/?mypy=latest&python=3.13&gist=c218144f43fd07fd2a2c93c142471b8f

> # assignment and the use here), but pyright supports this, presumably to provide features
> # like type-narrowing in the local scope. In conclusion, this should be `Literal[False]` or
> # `bool`.

Reviewer 2: ```suggestion
# assignment and the use here), but pyright and mypy support this.
# In conclusion, this could be `bool` but should  probably be `Literal[False]`.
```

c_instance = C()

> # Not that we would use this in static analysis, but for a more realistic example,, let's actually

Reviewer 2: ```suggestion
# Not that we would use this in static analysis, but for a more realistic example, let's actually
```
